pvlib/iotools/sam.py

```diff
@@ -0,0 +1,178 @@
+
+    """
+
+    def _is_leap_and_29Feb(s):
+        ''' Creates a mask to help remove Leap Years. Obtained from:
+            https://stackoverflow.com/questions/34966422/remove-leap-year-day-
+            from-pandas-dataframe/34966636
+        '''
+        return (s.index.year % 4 == 0) & \
+               ((s.index.year % 100 != 0) | (s.index.year % 400 == 0)) & \
+               (s.index.month == 2) & (s.index.day == 29)
+
+    def _averageSAMStyle(df, interval='60T', closed='right', label='right'):
+        ''' Averages subhourly data into hourly data in SAM's expected format.
+        '''
+        df = df.resample(interval, closed=closed, label=label).mean()
+        return df
+
+    def _fillYearSAMStyle(df, freq='60T'):
+        ''' Fills year
+        '''
+        # add zeros for the rest of the year
+        # add a timepoint at the end of the year
+        # apply correct TZ info (if applicable)
+        tzinfo = df.index.tzinfo
+        starttime = pd.to_datetime('%s-%s-%s %s:%s' % (df.index.year[0], 1, 1,
+                                                       0, 0)
+                                   ).tz_localize(tzinfo)
+        endtime = pd.to_datetime('%s-%s-%s %s:%s' % (df.index.year[-1], 12, 31,
+                                                     23, 60-int(freq[:-1]))
+                                 ).tz_localize(tzinfo)
+
+        df2 = _averageSAMStyle(df, freq)
+        df2.iloc[0] = 0  # set first datapt to zero to forward fill w zeros
+        df2.iloc[-1] = 0  # set last datapt to zero to forward fill w zeros
+        df2.loc[starttime] = 0
+        df2.loc[endtime] = 0
+        df2 = df2.resample(freq).ffill()
+        return df2
+
+    # Modify this to cut into different years. Right now handles partial year
+    # and sub-hourly interval.
+    if standardSAM:
+        filterdatesLeapYear = ~(_is_leap_and_29Feb(data))
+        data = data[filterdatesLeapYear]
+        data = _fillYearSAMStyle(data)
+
+    # metadata
+    latitude = metadata['latitude']
+    longitude = metadata['longitude']
+    elevation = metadata['elevation']
+    timezone_offset = metadata['TZ']
+    source = metadata['source']
+
+    # make a header
+    header = '\n'.join(['Source,Latitude,Longitude,Time Zone,Elevation',
+                        source + ',' + str(latitude) + ',' + str(longitude)
+                        + ',' + str(timezone_offset) + ',' +
+                        str(elevation)])+'\n'
+
+    savedata = pd.DataFrame({'Year': data.index.year,
+                             'Month': data.index.month,
+                             'Day': data.index.day,
+                             'Hour': data.index.hour})
+
+    if includeminute:
+        savedata['Minute'] = data.index.minute
+
+    windspeed = list(data.wind_speed)
+    temp_amb = list(data.temp_air)
+    savedata['Wspd'] = windspeed
+    savedata['Tdry'] = temp_amb
+
+    if 'dni' in data:
+        dni = list(data.dni)
+        savedata['DHI'] = dni
```

Comment:
I think you could just write savedata['X'] = data.X.values if this is to avoid index alignment.

Reply:
tx, pushed this change